### DISCUSSION TITLE
TaskId and JobId

### DISCUSSION THREAD

**Discussion by kirillsalykin** [maxcountryman/underway#66] [Q&A] [answered]
Hi,

I am still trying to implement `enqueue_multi` :)

But I ran into an question(s):
https://github.com/maxcountryman/underway/blob/main/src/job.rs#L907

```
   pub async fn cancel(&self) -> Result<bool> {
        let mut tx = self.queue.pool.begin().await?;
        let in_progress_tasks = sqlx::query_as!(
            InProgressTask,
            r#"
            select
              id as "id: TaskId",
              task_queue_name as "queue_name",
              input,
              retry_policy as "retry_policy: RetryPolicy",
              timeout,
              heartbeat,
              concurrency_key
            from underway.task
            where input->>'job_id' = $1
              and state = $2
            "#,
            self.id.to_string(),
            TaskState::Pending as TaskState
        )
        .fetch_all(&mut *tx)
        .await?;

        let mut cancelled = false;
        for in_progress_task in in_progress_tasks {
            if in_progress_task.mark_cancelled(&mut tx).await? {
                cancelled = true;
            }
        }
        tx.commit().await?;

        Ok(cancelled)
    }
}
```
What is `job_id` here? how it is different from the `task_id`? why do we need both?

I think it would make things a little simpler if there would be just the `id` which would identify a task.
Should I need a queue name in order to change task?
(sorry, I am again with this question/request :) )

1. you also probably want to have `select for update` to lock records, because they are about to be updated.
2. how it can be multiple tasks?
3. `mark_cancelled` probably should accept multiple `task`s as input, cancelling tasks one by one - not the most efficient approach
4. this makes me curious how do you see relation of tasks and queues? 
5. (I dont see how task can be in multiple queues?) what is the scenario(s) when tasks with the same `task_id` can belong to different `queues`?

**Comment by maxcountryman** [comment]
> What is job_id here? how it is different from the task_id? why do we need both?

It's an ID of the job, which is shared amongst all tasks.

> I think it would make things a little simpler if there would be just the id which would identify a task.

Jobs are composed of steps which are executed as independent tasks, so unfortunately there's no single task ID in that way.

I think part of what's tripping you up here is the fact that `underway.task` is implemented as a single table but the actual data model is not. Does it help if you think about it as a set of virtual tables? For an even more concrete example, take a look at `pg_boss`, which uses table partitions to literally create separate tables (this is something that we most likely do want to do in the future for performance reasons).

**Reply by maxcountryman** [reply, marked as answer]
This log output, from the `step` example, might also help you visualize how jobs work in practice:

```text
2024-11-16T16:51:00.717987Z  INFO process_next_task{queue.name="example-step" task.id="019335e2-8fcf-671e-858a-11aac1be2237"}:execute{job.id=019335e2-8fcf-3e1f-3b8b-5ff879f3bdd9 step=1 steps=3}: example_step: Starting computation with n = 42
2024-11-16T16:51:00.725515Z  INFO process_next_task{queue.name="example-step" task.id="019335e2-8fee-0a49-5407-ee3293301ed3"}:execute{job.id=019335e2-8fcf-3e1f-3b8b-5ff879f3bdd9 step=2 steps=3}: example_step: Squared value: 42^2 = 1764
2024-11-16T16:51:00.728310Z  INFO process_next_task{queue.name="example-step" task.id="019335e2-8ff5-e339-ca65-6ed2568cb0d2"}:execute{job.id=019335e2-8fcf-3e1f-3b8b-5ff879f3bdd9 step=3 steps=3}: example_step: Modulo 10 of 1764 is 4
```

Notice how there's one job ID but each step has a unique task ID. This allows you to identified all the steps a particular enqueue of a job took as well as e.g. cancel all related steps.

**Reply by kirillsalykin** [reply]
Ok, I think i understand it.
Huge thanks for explaination!

Would it make sense to make `job_id`, `step` and `steps` a separate column? wdyt?

**Reply by kirillsalykin** [reply]
another question on queue name and cancellation - it seems that `cancel` fn cancels tasks in any queue,
but AFAIK `job` = multiple `task`s in a specific queue, so I would expect that it should filter out tasks by queue name as well, am I being wrong?

**Reply by kirillsalykin** [reply]
(assuming each task can be in different queues and the same task can be in more than queue)

**Reply by maxcountryman** [reply]
> Would it make sense to make job_id, step and steps a separate column?

I don't believe so, no. 

Fundamentally the "job" concept is layered on top of the lower level primitives: task, queue, worker, and scheduler. Put another way, the lower level task queue is entirely unaware of the job concept.

That's a good and bad thing. It's less efficient to embed the job data model in JSON when querying. But the upside is the task queue implementation doesn't need to know anything about higher level concepts like jobs--they're entirely decoupled.

**Reply by kirillsalykin** [reply]
I think it would be nice to have it documented (it = if tasks of the job can be in multiple queues and if one task can be in multiple queues)

**Reply by maxcountryman** [reply]
> so I would expect that it should filter out tasks by queue name as well, am I being wrong?

You're correct. It should filter by queue name and task ID together.

**Reply by kirillsalykin** [reply]
> > Would it make sense to make job_id, step and steps a separate column?
> 
> I don't believe so, no.
> 
> Fundamentally the "job" concept is layered on top of the lower level primitives: task, queue, worker, and scheduler. Put another way, the lower level task queue is entirely unaware of the job concept.
> 
> That's a good and bad thing. It's less efficient to embed the job data model in JSON when querying. But the upside is the task queue implementation doesn't need to know anything about higher level concepts like jobs--they're entirely decoupled.

i think this is unrelated, having explicit columns - doesnt mean `task` should be aware of the job concept?
job_id is there, it is just implicit.

**Reply by kirillsalykin** [reply]
I am still curious which scenarios you have in mind for having tasks of the job in different queues?
or for one task being in two queues

**Reply by maxcountryman** [reply]
> I think it would be nice to have it documented (it = if tasks of the job can be in multiple queues and if one task can be in multiple queues)

Yes, I agree.

I think the important part is that the task table is really not a single table conceptually. It's not so much that tasks in practice will be copied onto multiple queues (although they may and we support this provided the same T::Input), it's that the single table is representing what is in practice multiple tables. So the only way to identify a unique task in a queue is with that queue's name + that task's ID. Maybe it makes sense to say it like this: there's one task table and tasks are unique for a given queue so in SQL we say the primary key is a composite of queue name + task ID. Apart from data model integrity, this also matters for query performance since the reality is we do only have a single table.

**Reply by maxcountryman** [reply]
> job_id is there, it is just implicit.

But I would argue that's the important part. The task is agnostic to the input type and that's exactly the value it's providing.

If we make it aware of job_id then we've started to couple tasks (a lower level concept) to jobs (a higher level concept). That isn't just a feeling of being a bit more yucky it also has an effect of making the system harder to reason about and iterate on.

Similarly if we have a separate job table now we need to teach other parts of the lower level system how to interface with that because jobs are no longer just an implementation of the task trait.

**Reply by maxcountryman** [reply]
> which scenarios you have in mind for having tasks of the job in different queues

There are logging and archival use cases but apart from that the fact that the worker takes a queue is meaningful here: it's possible to allocate more worker or scheduler resources to one queue over another which is another way to augment priority queuing.

**Reply by kirillsalykin** [reply]
> There are logging and archival use cases but apart from that the fact that the worker takes a queue is meaningful here: it's possible to allocate more worker or scheduler resources to one queue over another which is another way to augment priority queuing.

sorry, I am still not following why multiple queues might be needed here. having multiple workers can be achieved with one queue.

**Reply by kirillsalykin** [reply]
> Similarly if we have a separate job table now we need to teach other parts of the lower level system how to interface with that because jobs are no longer just an implementation of the task trait.

Which makes me think maybe we do need separate job table? For instance, `job` a high level concept build on top on `task` and `queue`, but currently `task` is aware about `queue`, which shouldn't be the case. (If I follow your logic correctly).

so I am thinking about something like this:
```
job
---
id
queue
steps


job_task
---
job_id
task_id
step
```

Just some random thoughts, wdyt?
@maxcountryman

**Reply by maxcountryman** [reply]
> having multiple workers can be achieved with one queue.

This doesn't allow you to have a "fast" queue and a not so fast queue.

I very much recommend exploring the `pg_boss` project which uses the same design.

**Reply by kirillsalykin** [reply]
I will look into pg_boss, thanks

**Reply by maxcountryman** [reply]
> Which makes me think maybe we do need separate job table?

No. I don't think so. As I mentioned, this would make things much more complex and require altering the lower level system. I'm not a fan of that approach.

**Reply by maxcountryman** [reply]
> currently task is aware about queue

I'm not following: queue and task are part of the same lower level system and have an explicit and specific relationship.

Taking this further the queue is explicitly designed to consider the task and more specifically to manage the task life cycle.

Likewise the task only exists in a queue so it makes sense for it to know about e.g. the queue name: it doesn't exist outside of the queue namespace.

The point is not that everything should be unaware of everything else but that the concepts should fit together in an intentional and specific way.

**Reply by kirillsalykin** [reply]
This makes me curious: why is it considered acceptable to have `job_queue` in the `tasks` table (as far as I know, `job_queue` doesn't logically belong to `tasks`), but not acceptable to include `job_id`?

**Reply by kirillsalykin** [reply]
Apologies, I may have misunderstood. I assumed that `job` is combination of `task` and `queue`.
So `task` and `queue` are separate concepts.

**Reply by maxcountryman** [reply]
> Apologies, I may have misunderstood. I assumed that job is combination of task and queue.

Hm, I don't think that's right: `Job` is an implementation of `Task`.

It pulls some tricks to allow you to define steps which are executed as tasks but if you set that aside for a moment it's a task like any other.

**Reply by maxcountryman** [reply]
> This makes me curious: why is it considered acceptable to have job_queue in the tasks table (as far as I know, job_queue doesn't logically belong to tasks), but not acceptable to include job_id?

Sorry I'm not following: there isn't a `job_queue`, is there? (Mobile at the moment so I apologize if I'm misremembering the schema here...I don't recall that being a column name or piece of state we track)

**Reply by kirillsalykin** [reply]
oh, my bad, it is `task_queue_name`
no questions :)

**Reply by maxcountryman** [reply]
Ah sorry, I should have realized from the context. I know I've said something to a similar effect before, but I think it's worth repeating: the data model is such the task table is representing a set of dynamic virtual tables where each table is a unique `task_queue_name`. Put another way, you could (maybe) do this by hand if you were to create tables named like `underway.{queue_name}_task`. If you had such tables, the primary key would be just ID. But note that, to query a task from a table like that you'd first need to know the name of the table, so you'd need to both the table name and the task ID.
